### PR TITLE
[PINOT-3203] Controller code changes for Low-level consumer segments

### PR DESCRIPTION
**Pull request by mcvsubbu**
Updated PinotLLCRealtimeSegmentManager to support closing an old segment in propertystore
and creating new ones in idealstate. Added tests for the new methods.

Added a tentative protocol file (actual protocol elements TBF) and an FSM implementation
for closing segments when multiple servers inform the controller about their offsets, and
the controller ensures that one of them commits the segment.

Added test cases to test the scenarios described in the design documents.

### CODE REVIEW

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ */
+package com.linkedin.pinot.common.protocols;
+
+import com.alibaba.fastjson.JSONObject;
+
+/*
+ * TODO Add unit tests for this after we finalize the protocol elements.
+ * TODO Finalize the protocol elements.
+ *
+ * This class encapsulates the segment finalization protocol used by the server and the controller for
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: I would change finalization to something else to avoid reader confusion with Java finalization.

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+   * MAX_SEGMENT_COMMIT_TIME_MS is the longest time (msecs) a server will take to complete building a segment and committing
+   * it  (via a SegmentCommit message) after the server has been notified that it is the committer.
+   */
+  public static final long MAX_SEGMENT_COMMIT_TIME_MS = 15000;
+
+  public enum ControllerResponseStatus {
+    NOT_SENT, // Never used by the controller. Only used locally by the server when it is not able to send a request
+    COMMIT,
+    HOLD,
+    CATCHUP,
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: Rename to `CATCH_UP`

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+    public Request(String segmentName, long offset, String instanceId) {
+      _segmentName = segmentName;
+      _instanceId = instanceId;
+      _offset = offset;
+    }
+
+    public abstract String getUrl(String hostPort);
+
+    protected String getUri(String method) {
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: Does method here refer to an HTTP request method (eg. GET, POST, DELETE, etc.) or something else?

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+   * send in a SegmentConsumedRequest with its current offset in kafka.
+   */
+  public static final long MAX_HOLD_TIME_MS = 3000;
+  /**
+   * MAX_SEGMENT_COMMIT_TIME_MS is the longest time (msecs) a server will take to complete building a segment and committing
+   * it  (via a SegmentCommit message) after the server has been notified that it is the committer.
+   */
+  public static final long MAX_SEGMENT_COMMIT_TIME_MS = 15000;
+
+  public enum ControllerResponseStatus {
```
Comment (jfim) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: Can we have a short description of each of these statuses? Eg. 
```
/** Server should discard segment */
DISCARD,
```

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 5]
```diff
@@ -198,6 +235,41 @@ public IdealState apply(IdealState idealState) {
...
+      @Override
+      public IdealState apply(IdealState idealState) {
+        return updateForNewRealtimeSegment(idealState, newInstances, oldSegmentNameStr, newSegmentNameStr);
+      }
+    }, RetryPolicies.exponentialBackoffRetryPolicy(5, 500L, 2.0f));
+  }
+
+  protected static IdealState updateForNewRealtimeSegment(IdealState idealState,
+      final List<String> newInstances, final String oldSegmentNameStr, final String newSegmentNameStr) {
+    if (oldSegmentNameStr != null) {
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: I'm confused by this. If we enter this code block, we update a code block-local `stateMap` but that variable doesn't escape that code block. Should the declaration of `stateMap` be pulled above the `if` statement?

Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: Ah I see what's happening, you're mutating the internal state of the ideal state. Would be better to do `idealState.setPartitionState`

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java (hunk 3 above)
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: It refers to one of MSG_TYPE_CONSUMED or MSG_TYPE_COMMIT

I will change it to msgType (or pathComponent)

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java (hunk 1 above)
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: Makes sense. Changed it to SegmentCompletionProtocol, and changed the other class names as well.

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java (hunk 2 above)
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: Done

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 6]
```diff
@@ -213,11 +285,167 @@ private IdealState addLLCRealtimeSegmentsInIdealState(final IdealState idealStat
...
+    idealStateEntries.put(newSegmentNameStr, newInstances);
+
+    List<String> paths = new ArrayList<>(2);
+    paths.add(oldZnodePath);
+    paths.add(newZnodePath);
+    List<ZNRecord> records = new ArrayList<>(2);
+    records.add(oldZnRecord);
+    records.add(newZnRecord);
+    /*
+     * Update zookeeper in two steps.
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: Good explanation of what's happening.

File: pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java (hunk 4 above)
Comment (mcvsubbu) [inline pinot-common/src/main/java/com/linkedin/pinot/common/protocols/SegmentFinalProtocol.java]: Done

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 7]
```diff
@@ -213,11 +285,167 @@ private IdealState addLLCRealtimeSegmentsInIdealState(final IdealState idealStat
...
+    }
+
+    Collections.sort(segmentNames, Collections.reverseOrder());
+
+    int curPartition = segmentNames.get(0).getPartitionId();
+    final int nSegments = segmentNames.size();
+
+    for (int i = 0; i < nSegments; i++) {
+      final LLCSegmentName segmentName = segmentNames.get(i);
+      if (segmentName.getPartitionId() == curPartition) {
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: This is kafka partition, right?

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+/**
+ * This is a singleton class in the controller that drives the state machines for segments that are in the
+ * committing stage.
+ */
+public class SegmentFinalizer {
+  // TODO Can we log using the segment name in the log message?
+  public static Logger LOGGER = LoggerFactory.getLogger(SegmentFinalizer.class);
+  private enum State {
+    HOLDING,          // the segment has started finalizing.
+      COMMITTER_DECIDED,
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java]: Weird indent.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+
+    public SegmentFinalProtocol.Response segmentConsumed(String instanceId, long offset) {
+      final long now = _segmentFinalizer.getCurrentTimeMs();
+      synchronized (this) {
+        LOGGER.info("Processing segmentConsumed({}, {})", instanceId, offset);
+        _commitStateMap.put(instanceId, offset);
+        SegmentFinalProtocol.Response response = null;
+        switch (_state) {
+          case HOLDING:
+            if (now > _startTime + MAX_TIME_TO_PICK_WINNER_MS || _commitStateMap.size() == _numReplicas) {
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java]: I'd add a short description of what's supposed to happen in each possible state. For example, in `HOLDING` state it could say something similar to "When a client reports its current offset, we want to wait up to `MAX_TIME` for `n` replicas to show up, at which point we choose the maximum offset and notify a replica."

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java (hunk 7 above)
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: Yes, I added a comment.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java (hunk 8 above)
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java]: Fixed

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java (hunk 9 above)
Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java]: Makes sense. I will add these comments.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+    }
+
+    public SegmentFinalProtocol.Response segmentConsumed(String instanceId, long offset) {
+      final long now = _segmentFinalizer.getCurrentTimeMs();
+      synchronized (this) {
+        LOGGER.info("Processing segmentConsumed({}, {})", instanceId, offset);
+        _commitStateMap.put(instanceId, offset);
+        SegmentFinalProtocol.Response response = null;
+        switch (_state) {
+          case HOLDING:
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java]: This state machine is quite complex, is there a way to simplify it?

I get the impression some states could be removed from it by using state that's been written to Helix/ZK in order to keep only the necessary parts for the segment completion phase to work. For example, if a client reports its offset but we've already committed the segment to the property store, there doesn't need to be a FSM state for this case; this would remove one state in the state machine as well as remove the need to keep the state machine in memory for an extended period of time.

If there's a diagram of the valid state transitions available on the GitHub wiki, it would also be nice to link it here to help with understanding this part of the protocol.

Comment (mcvsubbu) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentFinalizer.java]: It seems complex, perhaps due to lack of comments. 

I am adding the comments as we speak.

That aside, I did have the code that you mention outside the state machine at one point, but then it made sense to just run it through the FSM even in that case just because of the number of checks needed outside the FSM for this case.

The FSM does not stay for long time in memory. You can see the unit tests that confirm that the FSM gets out of memory.

Let me add some more comments, and let us see how it looks like.

**Comment by jfim** [comment]
I'm not sure I really like how this state machine is represented in code form and it's pretty hard to follow, but I don't have any concrete suggestions on how to improve it. It might be more readable if the states were classes, but some states reuse a fair bit of code, so it would still be convoluted to follow.

It's good that there's lots of logging, but it's still the kind of thing I hope won't break in the middle of the night in an obscure fashion. We should add a diagram that explains the various phases. With Graphviz installed, this should give you a basic diagram:

```
digraph segment_completion {
	HOLDING -> COMMITTER_NOTIFIED;
	HOLDING -> COMMITTER_DECIDED;
	COMMITTER_DECIDED -> COMMITTER_NOTIFIED;
	COMMITTER_DECIDED -> ABORTED;
	COMMITTER_NOTIFIED -> ABORTED;
	COMMITTER_UPLOADING -> ABORTED;
	COMMITTER_UPLOADING -> COMMITTING;
	COMMITTING -> COMMITTED;
	COMMITTING -> ABORTED;
	HOLDING -> ABORTED;
	COMMITTER_NOTIFIED -> COMMITTER_UPLOADING [ label = "Comment"];
}
```

You can turn the diagram into an image with `dot -O -Tpng test.dot`

I'll give this another pass when my brain is fully woken up later this morning.

File: pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java [hunk 11]
```diff
@@ -195,6 +199,24 @@ private synchronized SegmentCompletionFSM lookupOrCreateFsm(final LLCSegmentName
...
+   * constructor used.
+   *
+   * We kick off an FSM in the COMMITTED state (rare) when we find that PROPERTYSTORE already has the segment
+   * with the Status set to DONE.
+   *
+   * We kick off an FSM in the HOLDING state (typical) when a sementConsumed() message arrives from the
+   * first server we hear from.
+   *
+   * The FSM does not have a timer. It is clocked by the servers, which, typically, are retransmitting their
+   * segmentConsumed() message every so often (SegmentCompletionProtocol.MAX_HOLD_TOME_MS).
```
Comment (jfim) [inline pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/SegmentCompletionManager.java]: Typo in `MAX_HOLD_TIME_MS`

**Comment by jfim** [comment]
LGTM, thanks for reorganizing the code, it's a lot clearer what's happening now.
